### DISCUSSION TITLE
popen problems

### DISCUSSION THREAD

**Discussion by Godlike-Productions** [RomanKubiak/ctrlr#470] [Q&A] [answered]
OK, a little question out of left field.

Does anyone know why this won't run

```
function extractDirectory()

if isPanelReady() == false then -- check panel readiness
		    return
	    end
	
        extractPath = directoryToScanPath..ossep.."*.9kp"
        dosCmd = '"'..gzipPath..'gzip" -d -S 9kp -k -f "'..extractPath..'"'
        console ("dosCmd: "..dosCmd)
        tempOutput = io.popen(dosCmd)
end
```

The console output is

```
dosCmd: "C:\Program Files (x86)\GnuWin32\bin\gzip" -d -S 9kp -k -f "D:\Temp\*.9kp"
```

If I run
```
 "C:\Program Files (x86)\GnuWin32\bin\gzip" -d -S 9kp -k -f "D:\Temp\*.9kp" 
```
from a cmd window it works, and this exact code used to work in ctrlr yesterday, but is consistently not working today.

Happy to look at alternatives if someone knows how to gzip with raw LUA.  Sooo frustrating.  I had the tool working, and now it fails on this one step, which was working.

**Comment by Godlike-Productions** [comment]
This function works (though on files without spaces - the directory can have spaces, just not the file)

```
function testLoad()

metaKey = "metadata"
programKey = "program_id"
presetKey = "preset_id"
presetNameKey = "name"
algoNameKey = "program_name"

savePath = L(panel:getLabelComponent("filePathLabel"):getPropertyString("uiLabelText"))
gzipPath = L(panel:getLabelComponent("gzipPathLabel"):getPropertyString("uiLabelText"))

fileToConvert = utils.openFileWindow("Open Preset File to read (*.9kp)",File(savePath),"*.9kp",true)
fileToConvertPath = fileToConvert:getFullPathName()

if fileToConvert:existsAsFile() then
    console ("Preset File Exists: "..fileToConvertPath)
    tempFileName = fileToConvert:getFileName()
    tempOutput = io.popen('"'..gzipPath..'gzip" -d -S 9kp -k '..fileToConvertPath)
end --if

fileToParsePath = string.gsub(fileToConvertPath,".9kp","")
console (fileToParsePath)
 --fileToRead = utils.openFileWindow("Open json File to read (*.json)",File(savePath),"*.*",true)
fileToRead = File(fileToParsePath)

if fileToRead:existsAsFile() then

    console ("File Exists")
    str = fileToRead:loadFileAsString()
--    console (str)
    jStr = json.decode(str)
    
    if type(jStr) ~= "table" then
        console ("Not a Table")
        return
    else
        
    oldProgramID = jStr[metaKey][programKey]
    oldPresetID = jStr[metaKey][presetKey]
    presetName = jStr[metaKey][presetNameKey]
    algoName = jStr[metaKey][algoNameKey]

    oldBank = bit.band(bit.rshift(oldProgramID,16),0xff)
    oldAlgNum = bit.band(bit.rshift(oldProgramID,8), 0xff)
    oldPreset = bit.band(oldPresetID, 0xff)

    console (String("Algorithm Name: "..algoName))
    console (String("Preset Name: "..presetName))
    console (String("Old Bank: "..oldBank))
    console (String("Old Algorithm: "..oldAlgNum))
    console (String("Preset Number: "..oldPreset))

    end --if
    
end --if
end
```

**Comment by dnaldoog** [comment]
Would `tempOutput = io.popen('"'..gzipPath..'gzip" -d -S 9kp -k \"'..fileToConvertPath.."\"")` work?

**Comment by Godlike-Productions** [comment]
I'll give it a go.  Generally this type of stuff doesn't like the nested 

```"" ""```

I've got around it by adding the gzip directory to the windows PATH, so we avoid the 2 sets of quotes.  I read somewhere that 2 sets of quotes can cause some issues.

Also, I've put the command straight into the popen brackets, rather than referring via dosCmd.

It's working again for the moment.  

Setting gzip to PATH probably makes it easier for cross platform work because apparently OSX comes with gzip available straight from the terminal.  

This is a messy little utility

**Comment by Godlike-Productions** [comment, marked as answer]
This is what ended up working

```
tempOutput = io.popen('gzip -d -S 9kp -k "'..extractPath..'"')
```

but gzip is in the Windows PATH.  I'll try again with defining the command with the path in the command to see if I can make it work.

**Comment by dnaldoog** [comment]
That's great that you worked it out.

Another lua option that can be very useful is [loadstring](https://www.lua.org/pil/8.html) which executes a string as code. :)

**Reply by Godlike-Productions** [reply]
Thanks for that.  That one may prove useful.  If you find a pure lua library with no dependencies for gzip, let me know.  That would have avoided the popen altogether.

**Comment by dnaldoog** [comment]
Hi @Godlike-Productions 

Do you know about ZipFile class?
<pre style=color:dodgerblue>
			.def(constructor<const File &>())
			.def(constructor<InputStream *, bool>())
			.def(constructor<InputStream &>())
			.def(constructor<InputSource *>())
			.def("getNumEntries", &ZipFile::getNumEntries)
			.def("getEntry", (const ZipFile::ZipEntry *(ZipFile::*)(int) const noexcept) &ZipFile::getEntry)
			.def("getIndexOfFileName", &ZipFile::getIndexOfFileName)
			//.def("getEntry", (const ZipFile::ZipEntry *(ZipFile::*)(const String &) const noexcept) &ZipFile::getEntry)
			.def("sortEntriesByFilename", &ZipFile::sortEntriesByFilename)
			//.def("createStreamForEntry", (InputStream *(ZipFile::*)(int)) &ZipFile::createStreamForEntry, adopt(result))
			//.def("createStreamForEntry", (InputStream *(ZipFile::*)(const ZipFile::ZipEntry &)) &ZipFile::createStreamForEntry, adopt(result))
			.def("uncompressTo", &ZipFile::uncompressTo)
			.def("uncompressEntry", &ZipFile::uncompressEntry)
</pre>

----
[zipFile & SVG_0_1.zip](https://github.com/RomanKubiak/ctrlr/files/8809856/zipFile.SVG_0_1.zip)

![zipfilesshot](https://user-images.githubusercontent.com/76831683/171297217-cd73e00b-9a3e-43ad-9580-c2358c6b6156.png)

**Comment by Godlike-Productions** [comment]
Hmmmmm, interesting.  zip itself won't work, but it looks like the zipFile support gzip - https://docs.juce.com/master/classGZIPCompressorOutputStream.html

I'll have to have a look at this because OSX does not support io.popen and we need os.execute instead, but that's trickier.

I'll have a look at it when I get some time.

The other library I found is this one.

https://github.com/SafeteeWoW/LibDeflate

I think that would also work, as I'm pretty sure Deflate is the method used for gzip.

I specifically need gzip because the preset files for the Eventide H9000 are .json files that have been gzipped, so I need to uncompress, edit the .json and then recompress to allow the H9000 to load the file.
